highfidelity/hifi
# MyAvatar: Recenter when the head turns away from the hips

Compute the HMD facing moving average.  When the moving average diverges from the hips by more than 45 degrees, recenter the body.

Also, the follow code has been changed, instead of a follow velocity being passed to the CharacterController a desired target is passed.  The CharacterController homes toward it's target based on the time remaining.  Any follow deltas applied to move the avatar's position closer to it's target is stored and re-applied to the bodySensorMatrix.  This centralizes the moving/homing code to one place, the CharacterContoller.  The displacement each timestep is clamped, to avoid tunneling.

A new FollowHelper class was also introduced, it groups together the data and logic necessary to perform the re-centering/follow procedure.  This "hopefully" makes it easier to maintain.

Also, unit tests were added for new helper functions introduced to physics/BulletUtils.h

To test you must be wearing an HMD:

Once you are wearing an HMD, perform the following steps.

1. reset sensors using the "single quote" key.
2. look down and your avatar's toes/body should be pointing forward.
3. look to the right 90 degrees.  
4. Wait 3-4 seconds.
5. look down while maintaining your current head facing direction.  Your avatar toes should be aligned to your head facing.
